### ISSUE TITLE
Compiling on OS X is a mess

### ISSUE THREAD

**Issue by ghali** [otfried/ipe#33]
It is best to consider that OS X is just a unix machine. The best way for compiling ipe, until recently, has been to hack src/config.mak and treat OS X just like an ordinary unix machine. Just about every package available for Debian/Ubuntu/RHEL/etc.. is available on OS X by typing a command that is no longer than the equivalent one for Debian and co. It is very irritating to find that the last few releases of Ipe can all but not be compiled on OS X because the source files are inundated with halfhearted attempts at remedying one problem or another. Please make it possible for us to intentionally misspell "MACOS" and "Darwin" in the source files in one or a few places and henceforth OS X would be seen like a unix machine. Compilation is then entirely trivial. One just needs to have already installed Qt4 (<=7.1.9) or Qt5 (7.1.10), as well as texlive. These are massive packages that are best managed separately from Ipe. So whoever is maintaining the Ipe MacPort would help a lot by providing, in addition to or instead of, the Ipe MacPort with the full dependencies another Ipe MacPort that depends on neither texlive nor on Qt.

Hint: `pkg-config --cflags` and `pkg-config --libs` are available on OS X and work perfectly fine.

**Comment by otfried** [comment]
If you change "Darwin" in line 19 of "common.mak" to something else, compilation will proceed as if it's a normal Unix machine.   Compilation will then use "config.mak" and "pkg-config", not the special configuration file "macos.mak".  You may have to disable the special JPEG reading code in ipelib/ipebitmap_unix.cpp, too. If this doesn't work, you can file a real bug report instead of whining about "inundated with halfhearted attempts".

Apparently you think that running Ipe inside a macports or homebrew environment is the only real way of using a Mac.  Sorry, but you are a minority.  Most Ipe users on Mac just want something they can install and that will run, without having to install Macports, XQuartz, or a developer package for Qt.   My sources are such that I can compile such a bundle for normal Mac users.  And a normal Mac - without Macports - is really quite different from a Unix system.

If you want to treat your Mac as a Linux machine with a funny user interface, that's your choice.   Why would I have to go to any effort to support that? The last version compiles very easily on Mac if you just follow the installation instructions.  If you don't want to do that, you must be prepared to spend some effort.  Even though - as I said above - just disabling recognition of Darwin should mostly work.

pkg-config may work for you because you set it up right - in general it's a mess, because you never know if you get the native one, the one from homebrew, the one from macports, and very often something is broken.   And nobody stops any user from calling pkg-config from config.mak.  It's called "config.mak" because YOU are supposed to configure the compilation process.

**Comment by otfried** [comment]
I would be curious to know what your problems with compiling Ipe are.  Are you perhaps trying to compile the newest version using an outdated Macports recipe?

Compiling Ipe on Mac OS is easy.  As you say, you can easily install the few dependencies (libpng, libcairo, freetype, Qt), using macports, XQuartz, or otherwise.  

Once you have them, you set the correct flags in "macos.mak" (you are free to simply use pkg-config there, if it works on your system), and say "make".   

Ipe does not try to bundle or manage Qt, texlive, or any other dependency in any way.   I have no idea why you suggest that "These are massive packages that are best managed separately from Ipe" - of course, and Ipe  does not do any such thing.  Ipe simply links against them and works with whatever package you provided.

Perhaps your bug report is for the Macports recipe, not Ipe itself - but then it is entirely in the wrong place.   And really, one doesn't need a Macports recipe to compile Ipe, especially when one already has the necessary libraries installed through Macports.

**Comment by ghali** [comment]
I've been maintaining a recipe for compiling Ipe using MacPorts packages on OS X. The recipe remained rather stable for quite some time. I usually only upgraded when I had to. Still, I was pleasantly surprised to find that I needed to shrink the recipe. This went on until maybe 7.1.6. It got harder from that point and now I don't know how to compile either 7.1.9 or 7.1.10. Luckily I stowed away 7.1.8 from Mavericks (OS X 10.9), and that still works on Yosemite (10.10), but not from within emacs (nor from the OS X UI), only from the OS X Terminal (the shell). 

About the install location: I've been installing Ipe in /usr/local. It's worth mentioning that starting from El Capitan, Apple no longer makes it possible (or at least makes it harder) to use /usr/local for user-installed software. This will come up the day OS X folks have to upgrade.

With Ipe 7.1.10, the following gets over the first few difficulties.

In src/config.mak:

replace
CXX = g++
with
CXX = clang++

replace
LUA_PACKAGE   ?= lua5.2
with
LUA_PACKAGE   ?= lua

replace
JPEG_CFLAGS   ?=
JPEG_LIBS     ?= -ljpeg
with
JPEG_CFLAGS   ?= -I/opt/libjpeg-turbo/include
JPEG_LIBS     ?= -L/opt/libjpeg-turbo/lib -ljpeg

I am maintaining libjpeg-turbo separately (still under /opt, like other MacPorts packages). I couldn't install libjpeg using MacPorts because it clashes:

> sudo port install libjpeg-turbo
> Error: Unable to execute port: Can't install libjpeg-turbo because conflicting ports are active: jpeg

and jpeg is used by other MacPorts (ImageMagick may be one).

Now for a bit of ugliness. It may well be that pkg-config on Qt 5.4.2 is broken, but I found it necessary to replace:

QT_CFLAGS     ?= $(shell pkg-config --cflags Qt5Gui Qt5Widgets Qt5Core)
QT_LIBS       ?= $(shell pkg-config --libs Qt5Gui Qt5Widgets Qt5Core)

with

QT_CFLAGS     ?= -I/usr/local/Qt5.4.2/5.4/clang_64/lib/QtWidgets.framework/Versions/5/Headers \
-I/usr/local/Qt5.4.2/5.4/clang_64/lib/QtGui.framework/Versions/5/Headers \
-I/usr/local/Qt5.4.2/5.4/clang_64/lib/QtCore.framework/Versions/5/Headers \
-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Versions/A/Headers \
-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/AGL.framework/Headers \
-I/usr/local/Qt5.4.2/5.4/clang_64/mkspecs/macx-clang 
QT_LIBS       ?= -F/usr/local/Qt5.4.2/5.4/clang_64/lib -framework QtWidgets -framework QtGui -framework QtCore -framework DiskArbitration -framework IOKit -framework OpenGL -framework AGL

and that despite that I am duly setting PKG_CONFIG_PATH appropriately, as noted below.

For completeness let me record that pkg-config is not useful for Qt (5.4.2) because it uses:

-I/usr/local/Qt5.4.2/5.4/clang_64/include/QtWidgets 
-I/usr/local/Qt5.4.2/5.4/clang_64/include 
-I/usr/local/Qt5.4.2/5.4/clang_64/include/QtGui 
-I/usr/local/Qt5.4.2/5.4/clang_64/include 
-I/usr/local/Qt5.4.2/5.4/clang_64/include/QtCore 
-I/usr/local/Qt5.4.2/5.4/clang_64/include 

and this is clearly bogus. /usr/local/Qt5.4.2/5.4/clang_64/include doesn't even have these three subdirectories.

In src/common.mak

replace
  ifeq "$(UNAME)" "Darwin"
with anything other than Darwin
  ifeq "$(UNAME)" "Drown"

In

/usr/local/ipe-7.1.10/src/ipelib/ipebitmap_unix.cpp

replace the two instances of
`#ifdef __APPLE__`
with anything other than `__APPLE__`
`#ifdef __PPLE__`

Near line 188 replace:
soname      = -Wl,-soname,lib$1.so.$(IPEVERS)
with 
soname      = -Wl,-dylib_install_name,$(IPELIBDIRINFO)/lib$1.$(IPEVERS).dylib

BTW, thank you for upgrading to Qt5. Qt4 can now rest in peace. Since we're installing Qt without MacPorts, we need to set an environment variable in ~/.bash_profile:

export PKG_CONFIG_PATH="/usr/local/Qt5.4.2/5.4/clang_64/lib/pkgconfig:${PKG_CONFIG_PATH}"

With all the above, compilation of the source files containing `#include <QWidget>` does pass, but compilation bails out when compiling ipecanvas_qt.cpp with the error:

/usr/local/Qt5.4.2/5.4/clang_64/lib/QtWidgets.framework/Versions/5/Headers/qwidget.h:45:10: fatal error: 'QtWidgets/qsizepolicy.h' file not found
`#include <QtWidgets/qsizepolicy.h>`
         ^

My only hint (I haven't pursued this yet) is that Qt5 is different from Qt4 in that it requires an extra line in one's qmake/.pro file. That file must now have
QT += widgets
if one is using widgets
(or if you're trying to maintain code that works on both Qt4 and Qt5:
greaterThan(QT_MAJOR_VERSION, 4): QT += widgets
)
This may require an additional header directory include.

---

I understand that nothing in Ipe itself installs either Qt or TeX-Live. I was referring to the Ipe MacPort. Qmake would have been nice, if you had tolerated its idiosyncrasies. Installing via MacPorts is a second-best option. But there is really no need to bloat MacPorts with Qt (4 or 5), which also means that it would take hours to install Ipe. The slimmest solution is to download a precompiled Qt (/usr/local is a reasonable location). The MacPorts Ipe is improving. It used to depend on TeX-Live. Judging by the dry run (`sudo port install -y ipe`), it now only depends on qt4-mac (still for 7.1.9 or older), and TeX-Live can be installed/maintained separately.

As for OS X being just Linux with cooler graphics.. It's nice not to have to look online for the efforts of a hundred people who attempted and failed in installing hibernation on their Linux laptops. Plus the font hinting is pleasant.

For those trying to choose between MacPorts and homebrew, let me just mention that I found the difference to be somewhat like the difference between RHEL (/CentOS/..) and Fedora. MacPorts is conservative and stable. Homebrew is avant-gardish and adopts brittle changes. Thankfully MacPorts is maintained regularly.

**Comment by otfried** [comment]
I'm not sure why you insist on not using the MACOS setting in the Makefiles (common.mak).  It does only two things:  Use macos.mak instead of config.mak (but you could just copy your settings to that file instead), and set the right compilation and linking flags for OS X (you do that manually now anyway).

By default, "macos.mak" sets the IPEBUNDLE variable, which will nicely package all files into an Ipe.app bundle.  That's required if you want to run Ipe from the Finder, but you can of course unset the variable, install Ipe into /usr/local, and use Ipe only from the command line.

Since 7.1.8, Ipe no longer uses the turbojpeg API.  So you can simply link with -ljpeg using the  libjpeg from macports.  Turbojpeg is not needed.
In fact, on OS X Ipe would normally simply use the faster native function for decoding JPEG images.  That's what the lines

```
FRAMEWORKS      ?= /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks
JPEG_CFLAGS     ?= -I$(FRAMEWORKS)/CoreGraphics.framework/Versions/A/Headers/
JPEG_LIBS       ?= -framework ApplicationServices
```

in `macos.mak` are for.  But it seems you prefer to disable this in `ipebitmap_unix.cpp`.

The compilation error is inside a Qt 5 system header file - it's not finding `qsizepolicy.h`.  So it would seem some include directory is missing.   Some of the header files refer to other header files with a directory name, e.g.

`QWidget` does `#include <qwidget.h>`, but then `qwidget.h` does `#include <QtWidgets/qsizepolicy.h>`.   So the directory containing `qsizepolicy.h` should be named `QtWidgets` and must also be on the include search path.   Is there such a directory?  

You could compile one of the examples included with the Qt SDK and check what paths it uses.

Or just create the directories by making symbolic links:

```
cd /usr/local/Qt5.4.2/5.4/clang_64/lib/
ln -s . QtWidgets.framework/Versions/5/Headers/QtWidgets
ln -s . QtGui.framework/Versions/5/Headers/QtGui
ln -s . QtCore.framework/Versions/5/Headers/QtCore
```

**Comment by ghali** [comment]
All right, success. Thanks for your help.
It turns out that pkg-config is indeed broken on Qt—I asked on the Qt-interest mailing list. That remains the case in Qt 5.5.1. The following works on both Qt 5.4.2 and 5.5.1.

I needed to modify macos.mak. here are the differences:

diff macos-old.mak macos.mak
    11c11
    < X11LIBS      ?= $(IPESRCDIR)/../../dependencies
    ---
    > X11LIBS    = /opt/local
    21c21,23
    < FRAMEWORKS ?= /Developer/SDKs/MacOSX10.6.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks
    ---
    > FRAMEWORKS ?= /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks
    > 
    > 
    34c36
    < LUA_LIBS   ?= -L$(X11LIBS)/lib -llua53 -lm
    ---
    > LUA_LIBS   ?= -L$(X11LIBS)/lib -llua -lm
    36,40c38,49
    < QT_CFLAGS  ?= -I/Library/Frameworks/QtCore.framework/Versions/4/Headers \
    <              -I/Library/Frameworks/QtGui.framework/Versions/4/Headers
    < QT_LIBS    ?= -F/Library/Frameworks -L/Library/Frameworks \
    <              -framework QtCore -framework ApplicationServices \
    <              -framework QtGui -framework AppKit -framework Cocoa -lz -lm
    ---
    > QT_CFLAGS ?= -I/usr/local/Qt5.5.1/5.5/clang_64/lib/QtWidgets.framework/Versions/5/Headers \
    > -I/usr/local/Qt5.5.1/5.5/clang_64/lib/QtGui.framework/Versions/5/Headers \
    > -I/usr/local/Qt5.5.1/5.5/clang_64/lib/QtCore.framework/Versions/5/Headers \
    > -I/usr/local/Qt5.5.1/5.5/clang_64/lib \
    > -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/OpenGL.framework/Versions/A/Headers \
    > -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/AGL.framework/Headers \
    > -I/usr/local/Qt5.5.1/5.5/clang_64/mkspecs/macx-clang
    > QT_LIBS ?= -F/usr/local/Qt5.5.1/5.5/clang_64/lib \
    >  -framework QtWidgets -framework QtGui -framework QtCore \
    > -framework DiskArbitration -framework IOKit -framework OpenGL -framework AGL \
    > -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk \
    > -Wl,-rpath,/usr/local/Qt5.5.1/5.5/clang_64/lib

I needed the following soft links:

```
cd /usr/local/Qt5.5.1/5.5/clang_64/lib/
ln -s QtCore.framework/Versions/5/Headers ./QtCore
ln -s QtWidgets.framework/Versions/5/Headers ./QtWidgets
ln -s QtGui.framework/Versions/5/Headers ./QtGui
```

The recipe in macos.txt has two typos:

```
 cp ../fontmaps/mactex2011-fontmap.xml \
      ../../build/Ipe.app/Contents/Resources/fontmap.xml
```

should be
     cp ../fontmaps/mactex2011-fontmap.xml \
          ../build/Ipe.app/Contents/Resources/fontmap.xml

and
    Run the binary installer "qt-mac-opensource-4.8.2.dmg" for the Qt
    SDK for Mac from "http://download.qt.io/archive/qt/4.8/".
should be
    Run the binary installer "qt-opensource-mac-x64-clang-5.5.1.dmg" for the Qt
    SDK for Mac from "http://download.qt.io/archive/qt/5.5/5.5.1/".

I installed libpng, freetype, cairo, and lua through MacPorts:
https://www.macports.org/install.php
    sudo port install libpng
    sudo port install freetype
    sudo port install cairo
    sudo port install lua

**Comment by otfried** [comment]
Glad to hear it worked.

I compiled the last release with Qt 4 because Qt 5 no longer works with OS X 10.6.  (10.7 is still supported).  I'll see what I do in the next release.

**Comment by m7thon** [comment]
Just a couple of comments from my experience maintaining the MacPorts version of ipe:
- Using a package manager to compile ipe, or really any open source software, is very convenient, and in fact also really easy. This holds for linux, too, just that linux distributions have OSS package management built in.
- It is really a matter of taste which one you use. I tend to like MacPorts, but homebrew is also very popular. What I think does _not_ make much sense and is asking for trouble is mixing two package managers, i.e., installing dependencies via MacPorts but ipe itself via homebrew. If you install the dependencies via MacPorts, you may as well install ipe itself via MacPorts.
- The configuration above seems unnecessarily complicated. MacPorts really only sets the following environment variables, that's it:
  
  ```
  IPEQVORONOI=1 QHULL_CFLAGS=-I${prefix}/include/qhull
  IPEPREFIX=${prefix}
  ZLIB_CFLAGS="`pkg-config --cflags zlib`"
  ZLIB_LIBS="`pkg-config --libs zlib`"
  JPEG_LIBS="-ljpeg"
  PNG_CFLAGS="`pkg-config --cflags libpng`"
  PNG_LIBS="`pkg-config --libs libpng`"
  FREETYPE_CFLAGS="`pkg-config --cflags freetype2`"
  FREETYPE_LIBS="`pkg-config --libs freetype2`"
  CAIRO_CFLAGS="`pkg-config --cflags cairo`"
  CAIRO_LIBS="`pkg-config --libs cairo`"
  LUA_CFLAGS="`pkg-config --cflags lua`"
  LUA_LIBS="`pkg-config --libs lua`"
  QT_CFLAGS="`pkg-config --cflags QtGui QtCore`"
  QT_LIBS="`pkg-config --libs QtGui QtCore`"
  MOC=${qt_moc_cmd}
  ```
  
  Perhaps this can be used to improve the homebrew recipe as well.
- Setting `IPEBUNDLE=1` is _not_ required to _run_ ipe from the Finder as an Ipe.app, only to _package_ everything into an app bundle. That is not the same thing. For instance, MacPorts installs everything under `/opt/local/...`, but creates an `Ipe.app` under `/Applications/MacPorts/` that simply calls `/opt/local/bin/ipe`. You can then call ipe either from the Finder, or from the command line.
- Out of curiosity, I have been completely ignoring the fontmap.xml file (i.e., the MacPorts version of ipe has not used this for the last several years), and noone seems to have complained. Is that really needed for anything?

**Comment by m7thon** [comment]
Update: I just submitted an updated MacPorts Portfile for the current version 7.1.10, and I must say that I agree that compilation on OSX starting with this version is indeed a mess. I will file a separate issue with the problems that I encountered.

**Comment by otfried** [comment]
- A package manager is essential for Linux, because files are installed all over the system.  
  Without a package manager to keep track of which file belongs to which program, you would be 
  completely unable to uninstall software or to track dependencies.
  This doesn't apply to OS X, since applications are installed in self-contained bundles, at least
  if you follow Apple's architecture - so I think the need for a package manager is 
  somewhat overstated. 
- You are right, `fontmap.xml` has been completely useless for a while.  It is only needed when 
  PDF files do not contain embedded fonts (which is legal for the 14 standard PDF fonts, which 
  every PDF processing software must provide itself).   At least since Texlive 2013 the default
  setup of Texlive **does embed** even these standard fonts, so there is no more need for Ipe
  to find the fonts elsewhere.  I have removed the entire fontmap support from Ipe starting with
  the upcoming release.
- The OR's main issue was that he wanted to compile against Qt, not using the Qt in macports. 
  For some reason, Macports uses Qt 4 (which is officially no longer supported on El Capitan), 
  and it seems to use the X11 version of Qt (at least judging from the package name "qt4-x11"),
  which means you need to install an X-server just to use Ipe - this seems rather absurd to me.
  His problems seemed to mostly arise from the fact that Qt 5 for OS X doesn't seem to work 
  well with make-based compiles - I don't know what qmake does differently.

**Comment by otfried** [comment]
The next version of Ipe for OS X uses Cocoa directly, not Qt.   This means the install instructions for macports users will become very simple:

```
Quickstart using Macports
-------------------------

 * Ipe needs the libraries libpng, libcairo, freetype, and lua.  If
   you have macports (www.macports.org), you can install these easily
   by saying:

     sudo port install libpng lua cairo freetype

 * Compile Ipe by saying:

     cd src
     make       
     make app

 * Ipe is now available as a MacOS app in "build/Ipe.app".
```

Macports is severe overkill for this application, though.  Starting from a fresh install of El Capitan, I installed macports and then said `sudo port install libpng lua cairo freetype`.  Afterwards, my `/opt/local` tree used more than 800 MB, the compilation took ages.

When I compile libpng, lua, cairo, and freetype myself from sources (mostly very easy, just say configure & make), the Ipe dependencies fit in less than 10 MB.

**Comment by m7thon** [comment]
Hi Otfried. As you seem to be reading this issue thread, I will just add my current problems, and comments here:
1. Regarding MacPorts:
   - The MacPorts ipe version relies on Qt4 (not the x11 version), since there were issues with Qt5 last time I checked. The Qt4/5 status in MacPorts is tricky and changing.
   - My `/opt/local` tree uses several GB. It is the philosophy of MacPorts to reinstall a version of every dependency, even if it is already provided by Apple. The reason is that this tremendously eases support for different versions of OSX. After a while, pre-compiled versions of the main dependencies get cached, so installation remains fairly fast. Just for ipe, this is certainly overkill. But I have all my OSS (scientific python stack, octave, gcc, gimp, mpv (media player), latex, ...) managed by MacPorts, so for me this is not an issue.
   - If you rely on MacPorts to provide the dependencies for ipe, I don't see why you would not just install ipe via MacPorts as well.
   - For most users, a precompiled binary for OSX is probably the most convenient, though.
2. Current compilation issues:
   - The current `macos.mak` seems to be tailored to a very specific and non-standard setup, and is probably not very useful for general compilation on OSX (as noted by the OP). The `config.mak` file is very clear and much better suited, even on OSX.
   - The way you include CoreGraphics is AFAIK not correct, but I am not an expert. I think you should `#include <ApplicationServices/ApplicationServices.h>` instead of `#include <CoreGraphics.h>`. Then you don't need any include compiler directives, but only need to add `-framework ApplicationServices` to the library flags. (Please check this, though, since I am not sure.)
   - The `ipelets/lua/Makefile` causes my install to fail (at least with `IPEBUNDLE` unset). Perhaps the condition should be `ifdef IPEBUNDLE` instead of `ifdef MACOS`?
3. General comments:
   - Your above "very simple installation instructions" give me the chills. First of all, `sudo port install ipe` is certainly much simpler. But apart from this, why the heck would I want ipe bundled in an Ipe.app? I want to type `ipe <my_ipe_file>` and use it. If I wanted to go clicking around first, I would be using Windows...
   - Pl_eee_ase at least keep an option to just compile and install ipe to an IPEPREFIX as is possible now  . Note that such an install can still be wrapped by an Ipe.app for convenience, if desired (this is what the MacPort version does).
4. Thanks for all your great efforts! Ipe is a fantastic little piece of software :-)

**Comment by otfried** [comment]
Okay, let's reopen this as long as there is some discussion.
1. I didn't say that macports was bad and I understand why you use it.  I was just pointing out that just to compile Ipe, it's heavy overkill.   My own limited experience with macports is rather bad - after installing the Ipe dependencies, I wanted to install one more package - and the build immediately failed.
2. If Ipe uses Qt4 for Mac rather than for X11, why does http://www.macports.org/ports.php?by=name&substr=ipe then list `qt4-x11` as a dependence?  But I'm glad to hear it's not actually using this - in any case I don't want to worry about Qt anymore, and I'm not using it on OS X starting from the next release.
3. When I provide a Makefile for a platform, it should build a program according to the architecture of the platform.  For OS X, the architecture is clear:  applications are built as stand-alone bundles.  So the default compilation for OS X must built such a bundle.  It's as easy as this.

I don' t understand you concern that somehow having the Ipe binary in /opt/local/bin is "better" than having it in /Applications/Ipe.app/Contents/MacOS.   These are just conventions. 

As for "clicking around", I always use Ipe and its associated tools from the command line.  Simply add /Applications/Ipe.app/Contents/MacOS to your path, and you are all set.  (As for "I would be using Windows" - from my experience with bug reports on the different platforms, the average OS X user is even more clueless than the average Windows user.  Apple tries hard to keep their users uneducated.)
1. What is the exact error you get in ipelets/lua?  It seems to me that the default target is **all** and should do nothing.
2. The installation instructions are for somebody who wants to compile Ipe.  I assume people who  simply want to install through macports know how to type "sudo port install ipe" and don't need instructions for that.  So the instructions are meant for somebody who wants to compile Ipe without doing this - there are many reasons, e.g. wanting the newest version for which a portfile is not yet available.   The right thing to do in this situation is building a bundle.  It's self-contained, can be used immediately, moved around, installed in /Applications, and can be removed or replaced by a newer version at a keystroke.   It would be immensely foolish to somehow try to build into /usr/local (impossible on new OS X anyway) or into /opt/local (clobbering up the macports installation).
3. I believe macos.mak already says: If you don't want to build a bundle, undefine IPEBUNDLE and define all the variables from config.mak.    Essentially you are already doing this, except that you also decided to use libjpeg rather than Apple's native optimized jpeg code.

However, I'm not sure how well this is going to work starting with the next release, as my understanding is that Cocoa finds its resources (e.g. mainmenu.nib) inside the bundle which is defined by the path of the executable.  I'm sure you'll figure it out :-)   You could always compile against Qt and forgo the Cocoa-based improvements, I guess, in exchange for knowing that Ipe is where it belongs in /opt/local/bin...

**Comment by m7thon** [comment]
Hi there,
- MacPorts ipe depends on qt4-mac. The dependency is set based on an internal variable `os.platform`. If this is `darwin`, then qt4-mac is selected, otherwise qt4-x11. If you do `port info ipe` on your system, it will show the dependency correctly.
- The error with the Makefile in `ipelets/lua` is:
  
  ```
  make[2]: Entering directory `/opt/local/var/macports/build/_Users_mthon_dev_ports_graphics_ipe/ipe/work/ipe-7.1.10/src/ipelets/lua'  
  install -d /opt/local/var/macports/build/_Users_mthon_dev_ports_graphics_ipe/ipe/work/destroot../../../build/Ipe.app/Contents/Resources/ipelets  
  install: mkdir /opt/local/var/macports/build/_Users_mthon_dev_ports_graphics_ipe/ipe/work/destroot../../../build: Operation not permitted  
  make[2]: *** [install] Error 71
  ```
  
  Note that `INSTALL_ROOT='/opt/local/var/macports/build/_Users_mthon_dev_ports_graphics_ipe/ipe/work/destroot'`
  
  I think what is happening is that `IPELETDIR` gets overwritten to `$(BUILDDIR)/Ipe.app/Contents/Resources/ipelets` before it gets used in the `install` target, which is not what is intended.
- About the native jpeg code for the MacPorts version, I wasn't sure how to figure out the correct OSX SDK for the include directive, and figured it could not hurt to use libjpeg (or turbojpeg). This was simply the quickest solution. I later googled and think that you can simplify this by including `<ApplicationServices/ApplicationServices.h>` instead.
- Having a native Cocoa interface on OSX is great. If this means compiling everything into an app bundle, well, then so it is. As you say, there is nothing really wrong with this, and ipe can still be called from the command line. For the ipe application itself I guess I have to concede that this makes a lot of sense. However, ipe also comes with several command line programs, and in its core provides a shared library. I feel these are better installed in traditional unix paths (not in `/usr/local`, but in `IPEPREFIX`, i.e., wherever I want to put them).
  
  Note that building a self-contained Ipe.app bundle means that Ipe.app should not depend on non-system libraries installed somewhere on the system. Building an Ipe.app that has dependencies installed, well, "somewhere", is not a good idea. Then I would argue that it is better to install ipe into this "somewhere" (/usr/local, /opt/local, ~/local, whatever) as well and create an Ipe.app that calls ipe there.
- My comments on clicking around as in Windows were perhaps a little exaggerated and not entirely serious :-)
- The thing about self-compiling ipe and the instructions you provide is that this is probably something one should only attempt if you know what you are doing. Installing / compiling the ipe dependencies oneself is dangerous, since these will presumably end up somewhere in the users path and potentially interfere with other open source software. If these are updated, deleted, forgotten, ..., this can also break ipe. This is why I would really advise to use a package manager (and stick to one). Using e.g. MacPorts to install the dependencies, but then self-compile ipe is also problematic, since the dependencies can get updated in a way requiring a recompilation of ipe, or even accidentally get uninstalled.
  
  I understand that you want to provide easy to follow instructions for compiling ipe, but perhaps it is better to clearly state the requirements and leave it up to the user to figure out how to satisfy these.

**Comment by otfried** [comment]
"perhaps it is better to clearly state the requirements and leave it up to the user to figure out how to satisfy these"

But then I get bug reports like this one which tells me that "Just about every package available for Debian/Ubuntu/RHEL/etc.. is available on OS X by typing a command that is no longer than the equivalent one for Debian and co".

Perhaps the best would be not to provide instructions :-)

**Comment by otfried** [comment]
Thanks for the other pointers, I'll investigate.

**Comment by m7thon** [comment]
@ghali: "So whoever is maintaining the Ipe MacPort would help a lot by providing, in addition to or instead of, the Ipe MacPort with the full dependencies another Ipe MacPort that depends on neither texlive nor on Qt."

I am sorry, but even if I waned to, MacPorts will not allow such a port (using externally installed Qt), since it goes against the MacPorts idea of essentially providing all dependencies from within MacPorts (for maintainability reasons). Note that the MacPorts ipe port has never (explicitly) depended on MacPorts texlive, though, but perhaps some dependency chain brought this in, I don't know.

On modern hardware, I don't feel that this is really much of an issue anyway. Also, MacPorts now has a (currently not functioning) system to provide precompiled binaries, and with this in place, installation is basically as fast as `apt-get install ...` on linux.

Out of curiosity, I just moved my MacPorts to the side and did a complete reinstall from scratch, setting the default variants to `-x11 +no_x11 +quartz` in `/opt/local/etc/macports/variants.conf`, and did `sudo port install ipe`. This took a total of 52 minutes, of which 21 were spent building qt4-mac. The resulting `/opt/local` directory is 1.5 GB (which I can get down to 1 GB by removing stuff only required during the build). To be honest, I think this is not that bad.

**Comment by otfried** [comment]
_"This took a total of 52 minutes, ..  I think this is not that bad. "_

Maybe not - I didn't know about these cool flags.  Do they stop macports from adding all the weird dependencies to Cairo?

But honestly, why would _anyone_ want to install Ipe through macports?   Starting in the next release, the binary package for Ipe will be small and completely self-contained.  It no longer needs Qt, and contains compact builds of lua, libpng, libfreetype, and libcairo.

In the past, installing Ipe through macports was useful, because the binary Mac package often came out much after new sources, and until 7.1.9 the binary package assumed you had X11 installed (a huge dependency).  The binary package was also big because it contained the Qt monster, so by installing through macports you could share the macports Qt installation.

None of this will be true starting from the next release.  Binary Mac packages are released at the same time as the sources, they are small, and entirely self-contained.  

Some people may want to make changes to Ipe and therefore want to compile themselves.  But if you install through macports, you can't even do that.  Is there any reason why someone might prefer to install Ipe through macports?

**Comment by m7thon** [comment]
The `-x11` variant of cairo gets rid of the X11 dependencies (xrender, xorg-libXext, xorg-xcb-util). I have other software that really does require these, though, but it is useful if you can avoid X11 dependencies altogether. However, precompiled binaries, once available again, will only be provided for the default port variants.

> In the past, installing Ipe through macports was useful, because the binary Mac package often came out much after new sources, and until 7.1.9 the binary package assumed you had X11 installed (a huge dependency). The binary package was also big because it contained the Qt monster, so by installing through macports you could share the macports Qt installation."

Exactly. And back in 2010 as I recall there was no binary version for OSX (or very outdated). From then on, just maintaining the MacPorts port for ipe has not been that much effort.

> Starting in the next release, the binary package for Ipe will be small and completely self-contained. It no longer needs Qt, and contains compact builds of lua, libpng, libfreetype, and libcairo.
> 
> Binary Mac packages are released at the same time as the sources, they are small, and entirely self-contained.

Agreed, this will indeed be a great service to us OSX users, and at that point installing through MacPorts or homebrew becomes much less interesting. Will you be supporting older versions of OSX as well?

So now I am very curious about the next ipe release :-)

**Comment by otfried** [comment]
I'll build app bundles that should work on OS X 10,7 onwards.

It may take a few more days, as I'm also adding support for opentype fonts (through xelatex rather than pdflatex). I also need to figure out how to redesign the toolbar - Cocoa supports only one single toolbar. Any suggestions?

**Comment by m7thon** [comment]
If you have a look at the Preview.app, it does have two toolbars, one named "Markup Toolbar". Also, most OSX apps seem to allow customization of the toolbar. It makes a lot of sense if it is so restricted to choose your most important tools yourself. Otherwise, if you want to be really true to the Apple interface design, have a look at how Pages.app and Keynote.app do their interface (interestingly, though, I don't know anyone who actually uses these...). XCode uses a neat system with buttons that allow you to show and hide sidebars, and that is surprisingly useful.

Closer to the current ipe interface is the one of LibreOffice, but I have no idea what GUI toolkit they use.

Personally, I think there is nothing wrong with the current ipe interface, and it would be nice to keep it as cross-platform as possible.

**Comment by ghali** [comment]
On 23 December 2015 at 11:25, Michael Thon notifications@github.com wrote:

> I am sorry, but even if I waned to, MacPorts will not allow such a port
> (using externally installed Qt), since it goes against the MacPorts idea of
> essentially providing all dependencies from within MacPorts (for
> maintainability reasons).
> 
> That's unfortunate, because Ipe remained stable across the minor Qt
> updates, so someone who needed to track Qt closely could maintain a single
> Qt installation and use that for Ipe.

The news of hooking to Cocoa natively is great though, and it makes these
issues irrelevant. It will make the Ipe source code an object of study by
itself.

The worst was the dependence on X11 (or quartz), because the server alone
takes far too long to launch.

The best would be to add a feature to Ipe that makes it possible to refer
to a single diagram on a given page by a string reference, not by page
number. Then a document with many drawings can be accompanied with just one
file for figures, but it would remain possible to insert a diagram without
having to go through the latex file changing the page numbers accordingly.
I'm not sure if this is possible for those using the convenience of saving
Ipe figures in pdf format. Even if such a feature is added only to the
.xml/.ipe format, it would be very helpful, because then one could write a
brief makefile that builds the pdf file and the cross-references. How long
Ipe takes to launch would then be irrelevant, because the standard use-case
would be to launch Ipe once and leave it running. Right now having a large
number of windows is disorienting. But I'm digressing from the OS X issues.

My standard use-case on OS X is to point within emacs-dired to a figure
file and launch Ipe on it (either by opening it if it has the .ipe
extension, or by running the ipe executable on it via a shell escape).
Still, having an Ipe.app for quick sketching is a nice small luxury.

Note that the MacPorts ipe port has never (explicitly) depended on MacPorts

> texlive, though, but perhaps some dependency chain brought this in, I don't
> know.
> 
> The first time I ran `sudo port install ipe` (overnight), my recollection
> is that it brought texlive along—as well of course as Qt. That's history
> though.

MacPorts is a bit of a monster because it builds a parallel system to the
one provided by OS X, but its advantage (basically having the wonders of
both OS X and Linux running simultaneously without having to bother with
virtualization) more than outweigh the one-time waste in installation time
and carrying around the extra storage. The addition of texlive and Qt alone
significantly push that storage.

On modern hardware, I don't feel that this is really much of an issue

> anyway. Also, MacPorts now has a (currently not functioning) system to
> provide precompiled binaries, and with this in place, installation is
> basically as fast as apt-get install ... on linux.
> 
> Out of curiosity, I just moved my MacPorts to the side and did a complete
> reinstall from scratch, setting the default variants to -x11 +no_x11
> +quartz in /opt/local/etc/macports/variants.conf, and did sudo port
> install ipe. This took a total of 52 minutes, of which 21 were spent
> building qt4-mac. The resulting /opt/local directory is 1.5 GB (which I
> can get down to 1 GB by removing stuff only required during the build). To
> be honest, I think this is not that bad.

Agreed.

Sherif
